README


This archive contains the distribution Padre-Plugin-Ecliptic,
version 0.22:

  Padre plugin that provides Eclipse-like useful features

This software is copyright (c) 2010 by Ahmad M. Zawawi.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Changes for Padre-Plugin-Ecliptic

0.22   2010.10.29
	Removed Test::NeedsDisplay dependency (AZAWAWI)

0.21   2010.10.27
	t/00-load.t is magically injected by [@PadrePlugin] via [LoadTests] (AZAWAWI)

0.20   2010.10.23
	Switched to Dist::Zilla
	Fixed Compatibility with the latest Padre
	Share directory is now accessible via File::ShareDir

0.19  2009.12.17

	Fixed ticket:471 "Ecliptic Quick Module Access should use CPAN::SQLite::Search"

0.18  2009.8.24

	Removed stupid code from Ecliptic :)
	Reverting back to _sharedir until we get the Module::Build a sharedir
	Quick Fix is now in Padre 0.44 core
	Updated Arabic translation

0.17  2009.8.17

	Depends on Padre 0.43 in Build.PL
	Moved the following to Padre 0.43 core:
		Open resource dialog
		Quick menu access dialog
		Open in file browser action (integrated with directory browser)
	remove private _sharedir and use the one provided by the parent class (szabgab)
	Reuse Padre::Wx::Main help method (szabgab)
	Renamed to the new Padre 0.43 PO naming scheme
	POD fixes.
	Updated translations:
		German translation (Sewi++)
		French translation (jquelin++)

0.16  2009.8.2

	Updated translations:
		Arabic
		Brazilian portuguese (garu++)
		French (jquelin++)
		Dutch (ddn123456++)
		Japanese (ishigaki++)
	Implemented background searching in "Open resource"
	Added/Updated copyright/license for all the files
	"Open Resource" is truly a non-modal dialog and is smaller/easier to put on 
		the side
	Use plugin's icon on all dialogs...
	'Borrowed' some code to open files from Padre's Directory browser. 
	More UI final touches on "Open Resource"... more simpler...
	Handle Padre crash while calling a Padre action that should be called at that moment
		e.g. no documents...
	Display currently searched directory beside "Open Resource" dialog title
	Added an options popup menu for skip VCS and MANIFEST.SKIP
	"Open Resource" now respects MANIFEST.SKIP if it exists in project directory.
	You can now copy the file path.
	Selecting multiple files displays the number of files selected.
	Open Resource status label should display files paths all the time mimicking Eclipse's behavior.
	Changed "Open in Explorer" to "Open in File Browser" to be more less-win32 centric (El_Che++)
	Make sure the menu event is really valid code before calling it
	Quick Menu Access shall not crash anymore... A dream finally comes true.
	Depends on Padre 0.42 for the Padre Actions API
	Perl tidy and old code cleanup

0.15  2009.7.20

	- Ignore blib directory as well.
	- require Locale::Msgfmt 0.14
	- Ignore the blib directory as well (szabgab++).
	- Fixed ticket:456 "Ecliptic: quick menu access empty filter does not show 
		anything".
	- Unified behavior on quick menu, module and outline access.
	- Updated Arabic translation.
	- Keep the recently-opened 20 resources available when 
		you first open the "Open Resource" dialog. Productivity starts here.

0.14  2009.6.29

	- Only require Locale::Msgfmt when building the .mo files (RSN)

0.13  2009.6.27

	- Depends on Locale::Msgfmt to build the .mo files at install time (RSN, Ryan52++)

0.12  2009.6.24

	- Quick fix dialog has less size/position hard-coding. Takes a smaller font 
		than editor's line text height (i.e. works with CTRL +/-)
	- Ctrl+~ is out, Ctrl + Shift + 1 is now used for quick fix to be easier
		to use on uk/german keyboard shortcuts.
	- various POD fixes
	- POD shortcut update (e.g. Ctrl-3 => CTRL + 3)
	- a white background with a simple border like in Eclipse
	- Handled 'no filename' crash
	- Do nothing to exit the dialog when "No suggestion" is selected...

0.11  2009.6.21

	- Implemented "Quick Fix" popup list box which finds its list via 
		Padre::Document::SubClass->event_on_quick_fix.
	- Renamed List Key bindings into Quick Assist.

0.10  2009.6.10

	- Fixed "Open in Explorer" to work in KDE and GNOME
	- Updated translations:
		Arabic
		French (jquelin++)

0.09  2009.6.10

	- Implemented "Open in Explorer" action
	- POD fixes
	- Removing bad code. Padre::Current->main is now plugin->main
	- _T -> Wx::gettext. Alias++

0.08  2009.6.4

	- Implemented "List of key bindings" as detailed in ticket:379
	- Added translation:
		Polish (therek).

0.07  2009.5.31

	- [Open Resource] resource path is now shown in the status label.
	- Fixed a couple of GTK assertions
	- Any double-click event on list-box-es have now the same as pressing the OK button [garu++].
	- Avoid the debug assertion by ignoring the root item.

0.06  2009.5.31

	- Implemented "Quick Module Access" dialog.
	- Fixed 'Use of uninitialized value' on linux in
		"Open Resource" dialog (garu++)
	- Fixed the Kaboom bug in quick menu access in list box selection

0.05  2009.5.31

	- Implemented "Quick Outline Access" dialog.
	- Added translations:
		French (jquelin).
		Brazilian Portuguese (garu).
		Dutch (Dirk De Nijs).

0.04  2009.5.30

	- Implemented "Quick Menu Access" dialog.
	- Standard minimum size and centering for all dialogs.
	- Fixed a status message in "Open Resource" dialog.
	- Added Arabic translation.

0.03  2009.5.29

	- More POD fixes.
	- The "Open Resource" dialog has an ignore CVS/.svn/.git folders checkbox.

0.02  2009.5.28

	- Implemented "Open Resource" dialog.

0.01  2009.5.28

	- Initial release to CPAN.


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.30
    
    unless (eval "use Module::Build::Compat 0.02; 1" ) {
      print "This module requires Module::Build to install itself.\n";
      
      require ExtUtils::MakeMaker;
      my $yn = ExtUtils::MakeMaker::prompt
	('  Install Module::Build now from CPAN?', 'y');
      
      unless ($yn =~ /^y/i) {
	die " *** Cannot install without Module::Build.  Exiting ...\n";
      }
      
      require Cwd;
      require File::Spec;
      require CPAN;
      
      # Save this 'cause CPAN will chdir all over the place.
      my $cwd = Cwd::cwd();
      
      CPAN::Shell->install('Module::Build::Compat');
      CPAN::Shell->expand("Module", "Module::Build::Compat")->uptodate
	or die "Couldn't install Module::Build, giving up.\n";
      
      chdir $cwd or die "Cannot chdir() back to $cwd: $!";
    }
    eval "use Module::Build::Compat 0.02; 1" or die $@;
    use lib '_build/lib';
    Module::Build::Compat->run_build_pl(args => \@ARGV);
    require MyModuleBuilder;
    Module::Build::Compat->write_makefile(build_class => 'MyModuleBuilder');


lib/Padre/Plugin/Ecliptic.pm
package Padre::Plugin::Ecliptic;
BEGIN {
  $Padre::Plugin::Ecliptic::VERSION = '0.22';
}

# ABSTRACT: Padre plugin that provides Eclipse-like useful features

use strict;
use warnings;

# module imports
use Padre::Wx ();

# is a subclass of Padre::Plugin
use base 'Padre::Plugin';

#
# Returns the plugin name to Padre
#
sub plugin_name {
	return Wx::gettext("Ecliptic");
}

#
# This plugin is compatible with the following Padre plugin interfaces version
#
sub padre_interfaces {
	return 'Padre::Plugin' => 0.47;
}

#
# Returns the current share directory location
#
sub _sharedir {
	return Padre::Util::share('Ecliptic');
}

#
# plugin's real icon...
#
sub logo_icon {
	my ($self) = @_;

	my $icon = Wx::Icon->new;
	$icon->CopyFromBitmap( $self->plugin_icon );

	return $icon;
}

#
# plugin's bitmap not icon
#
sub plugin_icon {
	my $self = shift;

	# find resource path
	my $iconpath = File::Spec->catfile( $self->_sharedir, 'icons', 'ecliptic.png' );

	# create and return icon
	return Wx::Bitmap->new( $iconpath, Wx::wxBITMAP_TYPE_PNG );
}

#
# Called when the plugin is enabled
#
sub plugin_enable {
	my $self = shift;

	# Read the plugin configuration, and create it if it is not there
	my $config = $self->config_read;
	if ( not $config ) {

		# no configuration, let us write some defaults
		$config = {};
	}
	if ( not defined $config->{quick_menu_history} ) {
		$config->{quick_menu_history} = '';
	}

	# and write the plugin's configuration
	$self->config_write($config);
}

#
# called when Padre needs the plugin's menu
#
sub menu_plugins {
	my $self        = shift;
	my $main_window = shift;

	# Create a menu
	$self->{menu} = Wx::Menu->new;

	# Shows the "Quick Assist" dialog
	Wx::Event::EVT_MENU(
		$main_window,
		$self->{menu}->Append( -1, Wx::gettext("Quick Assist\tCtrl-Shift-L"), ),
		sub { $self->_show_quick_assist_dialog(); },
	);

	# Shows the "Quick Outline Access" dialog
	Wx::Event::EVT_MENU(
		$main_window,
		$self->{menu}->Append( -1, Wx::gettext("Quick Outline Access\tCtrl-4"), ),
		sub { $self->_show_quick_outline_access_dialog(); },
	);

	# Shows the "Quick Module Access" dialog
	Wx::Event::EVT_MENU(
		$main_window,
		$self->{menu}->Append( -1, Wx::gettext("Quick Module Access\tCtrl-5"), ),
		sub { $self->_show_quick_module_access_dialog(); },
	);

	#---------
	$self->{menu}->AppendSeparator;

	# the famous about menu item...
	Wx::Event::EVT_MENU(
		$main_window,
		$self->{menu}->Append( -1, Wx::gettext("About"), ),
		sub { $self->_show_about },
	);

	# Return our plugin with its label
	return ( $self->plugin_name => $self->{menu} );
}

#
# Shows the nice about dialog
#
sub _show_about {
	my ($main) = @_;

	my $about = Wx::AboutDialogInfo->new;
	$about->SetName("Padre::Plugin::Ecliptic");
	$about->SetDescription( Wx::gettext("Provides Eclipse-like useful features to Padre.\n") );
	$about->SetVersion($Padre::Plugin::Ecliptic::VERSION);
	Wx::AboutBox($about);

	return;
}

#
# Opens the "Quick Assist" dialog
#
sub _show_quick_assist_dialog {
	my $self = shift;

	#Create and show the dialog
	require Padre::Plugin::Ecliptic::QuickAssistDialog;
	my $dialog = Padre::Plugin::Ecliptic::QuickAssistDialog->new($self);
	$dialog->ShowModal();

	return;
}

#
# Opens the "Quick Outline Access" dialog
#
sub _show_quick_outline_access_dialog {
	my $self = shift;

	#Create and show the dialog
	require Padre::Plugin::Ecliptic::QuickOutlineAccessDialog;
	my $dialog = Padre::Plugin::Ecliptic::QuickOutlineAccessDialog->new($self);
	$dialog->ShowModal();

	return;
}

#
# Opens the "Quick Module Access" dialog
#
sub _show_quick_module_access_dialog {
	my $self = shift;

	#Create and show the dialog
	require Padre::Plugin::Ecliptic::QuickModuleAccessDialog;
	my $dialog = Padre::Plugin::Ecliptic::QuickModuleAccessDialog->new($self);
	$dialog->ShowModal();

	return;
}

1;



=pod

=head1 NAME

Padre::Plugin::Ecliptic - Padre plugin that provides Eclipse-like useful features

=head1 VERSION

version 0.22

=head1 SYNOPSIS

	1. After installation, run Padre.
	2. Make sure that it is enabled from 'Plugins\Plugin Manager".
	3. Once enabled, there should be a menu option called Plugins/Ecliptic.

=head1 DESCRIPTION

Once you enable this Plugin under Padre, you'll get a brand new menu with the
following options:

=head2 Quick Assist (Shortcut: Ctrl + Shift + L)

This opens a dialog with a list of current Padre actions/shortcuts. When
you hit the OK button, the selected Padre action will be performed.

=head2 Quick Outline Access (Shortcut: Ctrl + 4)

This opens a dialog where you can search for outline tree. When you hit the OK
button, the outline element in the outline tree will be selected.

=head2 Quick Module Access (Shortcut: Ctrl + 5)

This opens a dialog where you can search for a CPAN module. When you hit the OK
button, the selected module will be displayed in Padre's POD browser.

=head2 About

Shows a classic about dialog with this module's name and version.

=head1 Why the name Ecliptic?

I wanted a simple plugin name for including Eclipse-related killer features into
Padre. So i came up with Ecliptic and it turned out to be the orbit which the
Sun takes. And i love it!

=head1 AUTHOR

Ahmad M. Zawawi <ahmad.zawawi@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Ahmad M. Zawawi.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__



lib/Padre/Plugin/Ecliptic/QuickAssistDialog.pm
package Padre::Plugin::Ecliptic::QuickAssistDialog;
BEGIN {
  $Padre::Plugin::Ecliptic::QuickAssistDialog::VERSION = '0.22';
}

# ABSTRACT: Quick assist autocomplete dialog

use warnings;
use strict;

# module imports
use Padre::Wx ();

# is a subclass of Wx::Dialog
use base 'Wx::Dialog';

# accessors
use Class::XSAccessor accessors => {
	_plugin        => '_plugin',        # Plugin object
	_sizer         => '_sizer',         # window sizer
	_bindings_list => '_bindings_list', # key bindings list
};

# -- constructor
sub new {
	my ( $class, $plugin, %opt ) = @_;

	# create object
	my $self = $class->SUPER::new(
		$plugin->main,
		-1,
		Wx::gettext('Quick Assist'),
		Wx::wxDefaultPosition,
		Wx::wxDefaultSize,
		Wx::wxDEFAULT_FRAME_STYLE | Wx::wxTAB_TRAVERSAL,
	);

	$self->SetIcon(Wx::GetWxPerlIcon);
	$self->_plugin($plugin);

	# create dialog
	$self->_create;

	# Dialog's icon as is the same as plugin's
	$self->SetIcon( $plugin->logo_icon );

	return $self;
}


# -- event handler

#
# handler called when the ok button has been clicked.
#
sub _on_ok_button_clicked {
	my ($self) = @_;

	my $main = $self->_plugin->main;

	#Open the selected menu item if the user pressed OK
	my $selection = $self->_bindings_list->GetFirstSelected;
	if ( $selection != -1 ) {
		my $selected_menu_item = $self->_bindings_list->GetItem( $selection, 0 );
		if ($selected_menu_item) {
			my $event = Wx::CommandEvent->new( Wx::wxEVT_COMMAND_MENU_SELECTED,
				$selected_menu_item->GetId
			);
			$main->GetEventHandler->ProcessEvent($event);
		}
	}

	$self->Destroy;
}


# -- private methods

#
# create the dialog itself.
#
sub _create {
	my ($self) = @_;

	# create sizer that will host all controls
	my $sizer = Wx::BoxSizer->new(Wx::wxVERTICAL);
	$self->_sizer($sizer);

	# create the controls
	$self->_create_controls;
	$self->_create_buttons;

	# wrap everything in a vbox to add some padding
	$self->SetSizerAndFit($sizer);
	$sizer->SetSizeHints($self);

	# The dialog should be centered (more standard)
	$self->Centre;

	# focus on the search text box
	$self->_bindings_list->SetFocus;
}

#
# create the buttons pane.
#
sub _create_buttons {
	my ($self) = @_;
	my $sizer = $self->_sizer;

	my $butsizer = $self->CreateStdDialogButtonSizer( Wx::wxOK | Wx::wxCANCEL );
	$sizer->Add( $butsizer, 0, Wx::wxALL | Wx::wxEXPAND | Wx::wxALIGN_CENTER, 5 );
	Wx::Event::EVT_BUTTON( $self, Wx::wxID_OK, \&_on_ok_button_clicked );
}

#
# create controls in the dialog
#
sub _create_controls {
	my ($self) = @_;

	$self->_create_key_bindings_list();

	$self->_sizer->AddSpacer(10);
	$self->_sizer->Add( $self->_bindings_list, 0, Wx::wxALL | Wx::wxEXPAND, 2 );

	$self->_setup_events;

	return;
}

#
# Adds various events
#
sub _setup_events {
	my $self = shift;

	Wx::Event::EVT_LIST_ITEM_ACTIVATED(
		$self,
		$self->_bindings_list,
		sub {

			$self->_on_ok_button_clicked();
			$self->EndModal(0);

			return;
		}
	);

}

#
# Update matches list box from matched files list
#
sub _create_key_bindings_list {
	my $self = shift;

	$self->_bindings_list(
		Wx::ListView->new(
			$self,
			-1,
			Wx::wxDefaultPosition,
			[ 305, 300 ],
			Wx::wxLC_REPORT | Wx::wxLC_NO_HEADER | Wx::wxLC_SINGLE_SEL
		)
	);


	$self->_bindings_list->InsertColumn( 0, '' );
	$self->_bindings_list->InsertColumn( 1, '' );

	$self->_bindings_list->SetColumnWidth( 0, 180 );
	$self->_bindings_list->SetColumnWidth( 1, 100 );

	$self->_bindings_list->SetBackgroundColour( Wx::Colour->new( 255, 255, 225 ) );

	my $main     = $self->_plugin->main;
	my $menu_bar = $main->menu->wx;

	#walk the menu items tree
	sub walk_menu {
		my $menu = shift;

		my @menu_items = ();
		foreach my $menu_item ( $menu->GetMenuItems ) {
			my $sub_menu = $menu_item->GetSubMenu;
			if ($sub_menu) {
				push @menu_items, walk_menu($sub_menu);
			} elsif ( not $menu_item->IsSeparator ) {
				push @menu_items, $menu_item;
			}
		}

		return @menu_items;
	}

	my $menu_count = $menu_bar->GetMenuCount;
	my @menu_items = ();
	foreach my $menu_pos ( 0 .. $menu_count - 1 ) {
		my $main_menu       = $menu_bar->GetMenu($menu_pos);
		my $main_menu_label = $menu_bar->GetMenuLabel($menu_pos);
		push @menu_items, walk_menu($main_menu);
	}

	@menu_items = reverse sort { $a->GetLabel cmp $b->GetLabel } @menu_items;

	foreach my $menu_item (@menu_items) {
		my $menu_item_label    = $menu_item->GetLabel;
		my $menu_item_shortcut = $menu_item->GetText;

		if ( $menu_item_shortcut =~ /\t/ ) {
			$menu_item_shortcut =~ s/^.+?\t//;

			my $item;
			$item = Wx::ListItem->new();
			$item->SetText($menu_item_label);
			$item->SetData($menu_item);
			my $idx = $self->_bindings_list->InsertItem($item);

			$self->_bindings_list->SetItem( $idx, 1, $menu_item_shortcut );
		}
	}

	if ( $self->_bindings_list->GetItemCount() ) {
		$self->_bindings_list->Select( 0, 1 );
	}

	return;
}


1;

__END__
=pod

=head1 NAME

Padre::Plugin::Ecliptic::QuickAssistDialog - Quick assist autocomplete dialog

=head1 VERSION

version 0.22

=head1 AUTHOR

Ahmad M. Zawawi <ahmad.zawawi@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Ahmad M. Zawawi.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Padre/Plugin/Ecliptic/QuickMenuAccessDialog.pm
package Padre::Plugin::Ecliptic::QuickMenuAccessDialog;

use warnings;
use strict;

# package exports and version
our $VERSION = '0.08';
our @EXPORT_OK = ();

# module imports
use Padre::Wx ();
use Padre::Current ();
use Padre::Util   ('_T');

# is a subclass of Wx::Dialog
use base 'Wx::Dialog';

# accessors
use Class::XSAccessor accessors => {
	_plugin            => '_plugin',             # Plugin object
	_sizer             => '_sizer',              # window sizer
	_search_text       => '_search_text',	     # search text control
	_matches_list      => '_matches_list',	     # matches list
	_status_text       => '_status_text',        # status label
};

# -- constructor
sub new {
	my ($class, $plugin, %opt) = @_;

	# create object
	my $self = $class->SUPER::new(
		Padre::Current->main,
		-1,
		_T('Quick Menu Access'),
		Wx::wxDefaultPosition,
		Wx::wxDefaultSize,
		Wx::wxDEFAULT_FRAME_STYLE|Wx::wxTAB_TRAVERSAL,
	);

	$self->SetIcon( Wx::GetWxPerlIcon );
	$self->_plugin($plugin);

	# create dialog
	$self->_create;

	return $self;
}


# -- event handler

#
# handler called when the ok button has been clicked.
# 
sub _on_ok_button_clicked {
	my ($self) = @_;

	my $main = $self->_plugin->main;

	# Open the selected menu item if the user pressed OK
	my $selection = $self->_matches_list->GetSelection;
	my $selected_menu_item = $self->_matches_list->GetClientData($selection);
	if($selected_menu_item) {
		my $event = Wx::CommandEvent->new( Wx::wxEVT_COMMAND_MENU_SELECTED,  
			$selected_menu_item->GetId);
		$main->GetEventHandler->ProcessEvent( $event );
	}
	
	$self->Destroy;
}


# -- private methods

#
# create the dialog itself.
#
sub _create {
	my ($self) = @_;

	# create sizer that will host all controls
	my $sizer = Wx::BoxSizer->new( Wx::wxVERTICAL );
	$self->_sizer($sizer);

	# create the controls
	$self->_create_controls;
	$self->_create_buttons;

	# wrap everything in a vbox to add some padding
	$self->SetSizerAndFit($sizer);
	$sizer->SetSizeHints($self);

	# center the dialog
	$self->Centre;

	# focus on the search text box
	$self->_search_text->SetFocus;
}

#
# create the buttons pane.
#
sub _create_buttons {
	my ($self) = @_;
	my $sizer = $self->_sizer;

	my $butsizer = $self->CreateStdDialogButtonSizer(Wx::wxOK|Wx::wxCANCEL);
	$sizer->Add($butsizer, 0, Wx::wxALL|Wx::wxEXPAND|Wx::wxALIGN_CENTER, 5 );
	Wx::Event::EVT_BUTTON( $self, Wx::wxID_OK, \&_on_ok_button_clicked );
}

#
# create controls in the dialog
#
sub _create_controls {
	my ($self) = @_;

	# search textbox
	my $search_label = Wx::StaticText->new( $self, -1, 
		_T('&Type a menu item name to access:') );
	$self->_search_text( Wx::TextCtrl->new( $self, -1, '' ) );
	
	# matches result list
	my $matches_label = Wx::StaticText->new( $self, -1, 
		_T('&Matching Menu Items:') );
	$self->_matches_list( Wx::ListBox->new( $self, -1, [-1, -1], [400, 300], [], 
		Wx::wxLB_SINGLE ) );

	# Shows how many items are selected and information about what is selected
	$self->_status_text( Wx::StaticText->new( $self, -1, '' ) );
	
	$self->_sizer->AddSpacer(10);
	$self->_sizer->Add( $search_label, 0, Wx::wxALL|Wx::wxEXPAND, 2 );
	$self->_sizer->Add( $self->_search_text, 0, Wx::wxALL|Wx::wxEXPAND, 5 );
	$self->_sizer->Add( $matches_label, 0, Wx::wxALL|Wx::wxEXPAND, 2 );
	$self->_sizer->Add( $self->_matches_list, 0, Wx::wxALL|Wx::wxEXPAND, 2 );
	$self->_sizer->Add( $self->_status_text, 0, Wx::wxALL|Wx::wxEXPAND, 10 );

	$self->_setup_events;
	
	return;
}

#
# Adds various events
#
sub _setup_events {
	my $self = shift;
	
	Wx::Event::EVT_CHAR( $self->_search_text, sub {
		my $this  = shift;
		my $event = shift;
		my $code  = $event->GetKeyCode;

		if ( $code == Wx::WXK_DOWN ) {
			$self->_matches_list->SetFocus;
		}

		$event->Skip(1);		
	});

	Wx::Event::EVT_TEXT( $self, $self->_search_text, sub {

		$self->_update_matches_list_box;
		
		return;
	});
	
	Wx::Event::EVT_LISTBOX( $self, $self->_matches_list, sub {

		my $selection = $self->_matches_list->GetSelection;
		if($selection != Wx::wxNOT_FOUND) {
			$self->_status_text->SetLabel( 
				$self->_matches_list->GetString($selection));
		}
		
		return;
	});
	
	Wx::Event::EVT_LISTBOX_DCLICK( $self, $self->_matches_list, sub {
		$self->_on_ok_button_clicked();
		$self->EndModal(0);
	});
	
}

#
# Update matches list box from matched files list
#
sub _update_matches_list_box {
	my $self = shift;
	
	my $search_expr = $self->_search_text->GetValue;

	#quote the search string to make it safer
	$search_expr = quotemeta $search_expr;

	#Populate the list box now
	$self->_matches_list->Clear;
	my $pos = 0;
	
	my $main = $self->_plugin->main;
	my $menu_bar = $main->menu->wx;

	#File/New.../Perl 6 Script
	sub traverse_menu {
		my $menu = shift;
		
		my @menu_items = ();
		foreach my $menu_item ($menu->GetMenuItems) {
			my $sub_menu = $menu_item->GetSubMenu;
			if($sub_menu) {
				push @menu_items, traverse_menu($sub_menu);
			} elsif( not $menu_item->IsSeparator) {
				push @menu_items, $menu_item;
			}
		}
		
		return @menu_items;
	}
	
	my $menu_count = $menu_bar->GetMenuCount;
	my @menu_items = ();
	foreach my $menu_pos (0..$menu_count-1) {
		my $main_menu = $menu_bar->GetMenu($menu_pos);
		my $main_menu_label = $menu_bar->GetMenuLabel($menu_pos);
		push @menu_items, traverse_menu($main_menu);
	}
	@menu_items = sort { 
		$a->GetLabel cmp $b->GetLabel
	} @menu_items;
	foreach my $menu_item (@menu_items) {
		my $menu_item_label = $menu_item->GetLabel;
		if($menu_item_label =~ /$search_expr/i) {
			$self->_matches_list->Insert($menu_item_label, $pos, $menu_item);
			$pos++;
		}
	}
	if($pos > 0) {
		$self->_matches_list->Select(0);
		$self->_status_text->SetLabel("" . ($pos+1) . _T(' item(s) found'));
	} else {
		$self->_status_text->SetLabel(_T('No items found'));
	}
			
	return;
}


1;

lib/Padre/Plugin/Ecliptic/QuickModuleAccessDialog.pm
package Padre::Plugin::Ecliptic::QuickModuleAccessDialog;
BEGIN {
  $Padre::Plugin::Ecliptic::QuickModuleAccessDialog::VERSION = '0.22';
}

# ABSTRACT: Quick module autocomplete dialog

use warnings;
use strict;

# module imports
use Padre::Wx ();

# is a subclass of Wx::Dialog
use base 'Wx::Dialog';

# accessors
use Class::XSAccessor accessors => {
	_plugin       => '_plugin',       # Plugin object
	_sizer        => '_sizer',        # window sizer
	_search_text  => '_search_text',  # search text control
	_matches_list => '_matches_list', # matches list
	_status_text  => '_status_text',  # status label
	_modules      => '_modules',      # modules list
};

# -- constructor
sub new {
	my ( $class, $plugin, %opt ) = @_;

	# create object
	my $self = $class->SUPER::new(
		$plugin->main,
		-1,
		Wx::gettext('Quick Module Access'),
		Wx::wxDefaultPosition,
		Wx::wxDefaultSize,
		Wx::wxDEFAULT_FRAME_STYLE | Wx::wxTAB_TRAVERSAL,
	);

	$self->SetIcon(Wx::GetWxPerlIcon);
	$self->_plugin($plugin);

	# create dialog
	$self->_create;

	# Dialog's icon as is the same as plugin's
	$self->SetIcon( $plugin->logo_icon );

	return $self;
}


# -- event handler

#
# handler called when the ok button has been clicked.
#
sub _on_ok_button_clicked {
	my ($self) = @_;

	# Open the selected module in POD browser if the user pressed OK
	my $selection       = $self->_matches_list->GetSelection;
	my $selected_module = $self->_matches_list->GetClientData($selection);
	if ($selected_module) {
		Wx::Event::EVT_IDLE(
			$self,
			sub {

				#print "Show DocBrowser for '$selected_module'\n";
				Padre->ide->wx->main->help($selected_module);
				Wx::Event::EVT_IDLE( $self, undef );
			},
		);
	}

	$self->Destroy;

	return;
}


# -- private methods

#
# create the dialog itself.
#
sub _create {
	my ($self) = @_;

	# create sizer that will host all controls
	my $sizer = Wx::BoxSizer->new(Wx::wxVERTICAL);
	$self->_sizer($sizer);

	# create the controls
	$self->_create_controls;
	$self->_create_buttons;

	# wrap everything in a vbox to add some padding
	$self->SetSizerAndFit($sizer);
	$sizer->SetSizeHints($self);

	# center the dialog
	$self->Centre;

}

#
# create the buttons pane.
#
sub _create_buttons {
	my ($self) = @_;
	my $sizer = $self->_sizer;

	my $butsizer = $self->CreateStdDialogButtonSizer( Wx::wxOK | Wx::wxCANCEL );
	$sizer->Add( $butsizer, 0, Wx::wxALL | Wx::wxEXPAND | Wx::wxALIGN_CENTER, 5 );
	Wx::Event::EVT_BUTTON( $self, Wx::wxID_OK, \&_on_ok_button_clicked );
}

#
# create controls in the dialog
#
sub _create_controls {
	my ($self) = @_;

	# search textbox
	my $search_label = Wx::StaticText->new(
		$self, -1,
		Wx::gettext('&Type a module name to access:')
	);
	$self->_search_text( Wx::TextCtrl->new( $self, -1, '' ) );

	# matches result list
	my $matches_label = Wx::StaticText->new(
		$self, -1,
		Wx::gettext('&Matching modules:')
	);
	$self->_matches_list(
		Wx::ListBox->new(
			$self, -1, [ -1, -1 ], [ 400, 300 ], [],
			Wx::wxLB_SINGLE
		)
	);

	# Shows how many items are selected and information about what is selected
	$self->_status_text( Wx::StaticText->new( $self, -1, '' ) );

	$self->_sizer->AddSpacer(10);
	$self->_sizer->Add( $search_label,        0, Wx::wxALL | Wx::wxEXPAND, 2 );
	$self->_sizer->Add( $self->_search_text,  0, Wx::wxALL | Wx::wxEXPAND, 5 );
	$self->_sizer->Add( $matches_label,       0, Wx::wxALL | Wx::wxEXPAND, 2 );
	$self->_sizer->Add( $self->_matches_list, 0, Wx::wxALL | Wx::wxEXPAND, 2 );
	$self->_sizer->Add( $self->_status_text,  0, Wx::wxALL | Wx::wxEXPAND, 10 );

	$self->_setup_events;

	return;
}

#
# Adds various events
#
sub _setup_events {
	my $self = shift;

	Wx::Event::EVT_CHAR(
		$self->_search_text,
		sub {
			my $this  = shift;
			my $event = shift;
			my $code  = $event->GetKeyCode;

			if ( $code == Wx::WXK_DOWN ) {
				$self->_matches_list->SetFocus;
			}

			$event->Skip(1);
		}
	);

	Wx::Event::EVT_TEXT(
		$self,
		$self->_search_text,
		sub {

			$self->_update_matches_list_box;

			return;
		}
	);

	Wx::Event::EVT_LISTBOX(
		$self,
		$self->_matches_list,
		sub {

			my $selection = $self->_matches_list->GetSelection;
			if ( $selection != Wx::wxNOT_FOUND ) {
				$self->_status_text->SetLabel( $self->_matches_list->GetString($selection) );
			}

			return;
		}
	);

	Wx::Event::EVT_LISTBOX_DCLICK(
		$self,
		$self->_matches_list,
		sub {
			$self->_on_ok_button_clicked();
			$self->EndModal(0);
		}
	);

	Wx::Event::EVT_IDLE(
		$self,
		sub {

			# update matches list
			$self->_update_matches_list_box;

			# focus on the search text box
			$self->_search_text->SetFocus;

			# unregister from idle event
			Wx::Event::EVT_IDLE( $self, undef );
		}
	);
}

#
# Update matches list box from matched files list
#
sub _update_matches_list_box {
	my $self = shift;

	my $search_expr = $self->_search_text->GetValue;

	#quote the search string to make it safer
	$search_expr = quotemeta $search_expr;

	#Populate the list box now
	$self->_matches_list->Clear;
	my $pos = 0;

	unless ( $self->_modules ) {
		$self->_status_text->SetLabel( Wx::gettext("Reading modules. Please wait...") );
		my %seen;
		for my $path (@INC) {
			for my $file ( File::Find::Rule->name('*.pm')->in($path) ) {
				my $module = substr( $file, length($path) + 1 );
				$module =~ s/.pm$//;
				$module =~ s{[\\/]}{::}g;
				$seen{$module}++;
			}
		}
		my @modules = sort keys %seen;
		$self->_modules( \@modules );
		$self->_status_text->SetLabel( Wx::gettext("Finished Searching") );
	}

	foreach my $module ( @{ $self->_modules } ) {
		if ( $module =~ /$search_expr/i ) {
			$self->_matches_list->Insert( $module, $pos, $module );
			$pos++;
		}
	}

	if ( $pos > 0 ) {
		$self->_matches_list->Select(0);
		$self->_status_text->SetLabel( "" . ( $pos + 1 ) . Wx::gettext(' item(s) found') );
	} else {
		$self->_status_text->SetLabel( Wx::gettext('No items found') );
	}

	return;
}

1;

__END__
=pod

=head1 NAME

Padre::Plugin::Ecliptic::QuickModuleAccessDialog - Quick module autocomplete dialog

=head1 VERSION

version 0.22

=head1 AUTHOR

Ahmad M. Zawawi <ahmad.zawawi@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Ahmad M. Zawawi.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  'build_requires' => {
    'Locale::Msgfmt' => '0.14',
    'Module::Build' => '0.3601',
    'Test::More' => '0'
  },
  'configure_requires' => {
    'Module::Build' => '0.3601'
  },
  'dist_abstract' => 'Padre plugin that provides Eclipse-like useful features',
  'dist_author' => [
    'Ahmad M. Zawawi <ahmad.zawawi@gmail.com>'
  ],
  'dist_name' => 'Padre-Plugin-Ecliptic',
  'dist_version' => '0.22',
  'license' => 'perl',
  'module_name' => 'Padre::Plugin::Ecliptic',
  'recommends' => {},
  'recursive_test_files' => 1,
  'requires' => {
    'Class::XSAccessor' => '1.02',
    'File::Which' => '0.05',
    'Padre' => '0.72'
  },
  'script_files' => [],
  'share_dir' => {
    'dist' => 'share'
  }
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
